src/Autocomplete/src/AutocompleteResultsExecutor.php

```diff
@@ -46,14 +50,57 @@ public function fetchResults(EntityAutocompleterInterface $autocompleter, string
 
         $entities = $queryBuilder->getQuery()->execute();
 
-        $results = [];
+        $options = [];
+        $optgroupLabels = [];
+
+        if (null !== $groupBy = $autocompleter->getFormOption('group_by')) {
+            if (\is_string($groupBy)) {
+                $groupBy = new PropertyPath($groupBy);
+            }
+
+            if ($groupBy instanceof PropertyPathInterface) {
+                $accessor = PropertyAccess::createPropertyAccessor();
+                $groupBy = function ($choice) use ($accessor, $groupBy) {
+                    try {
+                        return $accessor->getValue($choice, $groupBy);
+                    } catch (UnexpectedTypeException) {
+                        return null;
+                    }
+                };
+            }
+
+            if (\is_callable($groupBy)) {
+                foreach ($entities as $entity) {
+                    $option = [
+                        'value' => $autocompleter->getValue($entity),
+                        'text' => $autocompleter->getLabel($entity),
+                    ];
+
+                    $groupLabels = $groupBy($entity, $option['value'], $option['text']);
+
+                    if (null !== $groupLabels) {
+                        $groupLabels = \is_array($groupLabels) ? array_map('strval', $groupLabels) : [(string) $groupLabels];
+                        $option['group_by'] = $groupLabels;
+                        $optgroupLabels = array_merge($optgroupLabels, $groupLabels);
+                    }
+
+                    $options[] = $option;
+                }
+
+                return [
+                    'options' => $options,
+                    'optgroups' => array_map(fn (string $label) => ['value' => $label, 'label' => $label], array_unique($optgroupLabels)),
+                ];
```

Comment:
The additions to this file as the riskiest part of this PR form a perspective if missing some edge cases and introducing bugs. Did you copy/adapt your code here form somewhere in core that handles the `group_by` option?

Reply:
Yes, I was inspired by what is done in the form component
https://github.com/symfony/form/blob/6.1/ChoiceList/Factory/PropertyAccessDecorator.php#L168
https://github.com/symfony/form/blob/6.1/ChoiceList/Factory/DefaultChoiceListFactory.php#L95